gluster_exporter/glusterinterface/gluster_interface.go

```diff
@@ -0,0 +1,323 @@
+}
+
+// Volume describes gluster volume
+type Volume struct {
+	XMLName   xml.Name `xml:"volume"`
+	VolName   string   `xml:"volName"`
+	NodeCount int      `xml:"nodeCount"`
+	Node      []Node   `xml:"node"`
+	Status    string
+}
+
+// Node describes gluster brick node
+type Node struct {
+	XMLName  xml.Name `xml:"node"`
+	Hostname string   `xml:"hostname"`
+	Path     string   `xml:"path"`
+	PeerID   string   `xml:"peerid"`
+	Status   string   `xml:"status"`
+	Port     string   `xml:"port"`
+	Ports    struct {
+		TCP  string `xml:"tcp"`
+		RDMA string `xml:"rdma"`
+	} `xml:"ports"`
+	Pid string `xml:"pid"`
+}
+
+// GD1 enables users to interact with gd1 version
+type GD1 struct {
+}
+
+// ExecuteCmd enables to execute system cmds and returns stdout, err
+func ExecuteCmd(cmd string) ([]byte, error) {
+	cmdfields := strings.Fields(cmd)
+	cmdstr := cmdfields[0]
+	args := cmdfields[1:]
+	fmt.Println("executing", cmdstr, args)
+	out, err := exec.Command(cmdstr, args...).Output()
+	return out, err
+}
+
+// GetOnlinePeers gets you the online gluster peers, look for Peer struct
+func (g GD1) GetOnlinePeers() ([]Peer, error) {
+	cmd := "gluster peer status --xml"
+	out, err := ExecuteCmd(cmd)
+	if err != nil {
+		return nil, err
+	}
+	var result peerStatus
+	err = xml.Unmarshal(out, &result)
+	var onlinePeers []Peer
+	for _, peer := range result.Peers {
+		if peer.Connected == 1 {
+			onlinePeers = append(onlinePeers, peer)
+		}
+	}
+	return onlinePeers, err
+}
+
+// GetHealInfo gets gluster vol heal in []HealEntries
+func (g GD1) GetHealInfo(vol string) ([]HealEntries, error) {
+	cmd := fmt.Sprintf("gluster vol heal %s info --xml", vol)
+	out, err := ExecuteCmd(cmd)
+	if err != nil {
+		return nil, err
+	}
+	var healop healBricks
+	err = xml.Unmarshal(out, &healop)
+	return healop.Healentries, err
+}
+
+// GetVolStatus gets the volume status if given vol name,
+// otherwise gets all vols status
+func (g GD1) GetVolStatus(vol string) ([]Volume, error) {
+
+	cmd := fmt.Sprintf("gluster vol status %s --xml", vol)
+
+	out, err := ExecuteCmd(cmd)
+	if err != nil {
+		return nil, err
+	}
+	var volstatus volumeStatus
+	err = xml.Unmarshal(out, &volstatus)
+	if err == nil {
+		Volumes := volstatus.Volumes
+		for index := 0; index < len(Volumes); index++ {
+			// get brick status
+			vol := &Volumes[index]
+			for bindex := 0; bindex < len(vol.Node); bindex++ {
+				brick := &vol.Node[bindex]
+				if brick.Status == "1" {
+					vol.Status = "Started"
+					break
+				}
+
+			}
+		}
+	}
+	return volstatus.Volumes, err
+}
+
+// GD2 is struct to interact with Glusterd2 using REST API
+type GD2 struct {
+	Host string
+	Port string
+}
+
+// MakeGD2 returns created gd2 instance, given host and port
+func MakeGD2(host string, port string) GD2 {
+	gd2 := GD2{host, port}
+	return gd2
+}
+
+// GetOnlinePeers gets online peers from Glusterd2 using REST
+func (g GD2) GetOnlinePeers() ([]Peer, error) {
+	url := "http://" + g.Host + ":" + g.Port + "/v1/peers"
```

Comment:
even http/https is configurable. We can use `restclient` package from gd2 https://github.com/gluster/glusterd2/tree/master/pkg/restclient